raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java

```diff
@@ -92,4 +102,24 @@ public static Path createTempFile(Path logDir, OffsetAndEpoch snapshotId) throws
 
         return Optional.of(new SnapshotPath(path, new OffsetAndEpoch(endOffset, epoch), partial));
     }
+
+    /**
+     * Delete the snapshot from the filesystem, we will firstly tried to rename snapshot file to
+     * ${file}.deleted, or delete the file directly if its already renamed.
+     */
+    public static boolean deleteSnapshotIfExists(Path logDir, OffsetAndEpoch snapshotId) {
+        Path path = Snapshots.snapshotPath(logDir, snapshotId);
+        Path destination = Snapshots.deleteRename(path, snapshotId);
+        try {
+            // rename before deleting if target file is not renamed
+            if (Files.exists(path) || Files.notExists(destination)) {
+                Utils.atomicMoveWithFallback(path, destination);
+            }
+            return Files.deleteIfExists(destination);
```

Comment:
Hmm. How about?

```suggestion
            return Files.deleteIfExists(path) | Files.deleteIfExists(destination);
```

And maybe rename the variables `path` and `destination` to something like `immutablePath` and `deletingPath` respectively.

Reply:
This looks better, I change the code and the annotations.